### PR TITLE
feat(revit): adds fallback option to receive displayable speckle objects as directshapes

### LINKED ISSUE #2468: bug(Revit): Receive objects as Direct Mesh not working as expected
When importing as DirectShapes some elements are missing.

- nested elements
- groups
- trees
- furniture
- site elements

### Test Model 
Revit House loaded with Speckle Connector as DirectShape is not visually the same as in Speckle viewer. 

Test model: rac_basic_sample_project.rvt
Commit: https://speckle.xyz/streams/9ff64f51e6/commits/d925c44f18

Missing objects in the test model: 

- [planting](https://speckle.xyz/streams/9ff64f51e6/objects/09ec296fd9eba79efaee1adc5e638eeb) - family: RPC Tree
- [PV panels](https://speckle.xyz/streams/9ff64f51e6/objects/671173efcc176534b76e96f963cf2710) - nested groups with electrical equipment, family: Photovoltaic-Panel-…, type: SunModule SW 245 Silver Mono - 10 Deg. Angle
- [site](https://speckle.xyz/streams/9ff64f51e6/objects/f6065e382d02928e45134350045d95cd) - family: M_Wind Power Generator…
- [entourage](https://speckle.xyz/streams/9ff64f51e6/objects/0a9c0aef0370e652410d2c2ae82feea4) - family: RPC Male…
- [furniture](https://speckle.xyz/streams/9ff64f51e6/objects/4d3a29fce7d4c86627c6e4f8a77da58e) - family: Seating - Artemis …
- [generic](https://speckle.xyz/streams/9ff64f51e6/objects/e3fd9a4d9522a173452fda14249972c2) - fireplace…
- all windows missing
- curtain walls received as walls

Untitled

### Desired Behavior:
- when receiving a model with this option enabled, all the meshes also visible in the viewer should be created in Revit with the right category associated to them
- any received model has to match its appearance in Speckle viewer and connector preview

### Extra:
- Rename the option to "Receive objects as DirectShape"

### LINKED ISSUE #2777: DirectShape fallback on receive
When receiving in Revit, we should always fallback to a DirectShape conversion (with a matching category) in case:

- it's receiving native objects that are unsupported (`CanConvert` does not return true)
- the conversion fails or is known to be buggy (eg roofs) NOTE: instead of returning null when the conversion fails, we need to call this fallback method
- ~LATER elements with no matching family & type (to be investigated as an option for the type-matching dialog)~

### PR DESCRIPTION
**Pull request by clairekuang**
## Description & motivation

This feature includes the following:
- Fixes the `directshape receive setting` so that all commit objects are received as directshapes created from their displayValue geometry. This should look exactly as the commit does in the viewer.
- Changes the way traversal of hosted elements happens. It used to be at the converter level, but it is now the connector level.
- Modifies the DirectShape receive setting to now have 3 options: "Always" (i.e. the old DirectMesh receive with all the new fixes), "On Fail" (The fallback DS behaviour, when some conversion goes wrong), "Never" (will try to convert and if it fails will return the error with no fallback attempt)

Closes #2777
Closes #2468

## Changes:

- Revit Connector and Converter
- 
## To-do before merge:

- [ ] More thorough testing of Revit receive workflows from different connectors, especially nested/hosted elements
- [x] (opt) set fallback behavior as a setting, rather than use it by default

## Validation of changes:

Tested receiving large revit commits with instances using the directshape setting.

### CODE REVIEW

File: ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs [hunk 1]
```diff
@@ -202,68 +241,65 @@ private void DeleteObjects(IReceivedObjectIdMap<Base, Element> previousObjects,
...
-
           var s = new CancellationTokenSource();
           DispatcherTimer.RunOnce(() => s.Cancel(), TimeSpan.FromMilliseconds(10));
           Dispatcher.UIThread.MainLoop(s.Token);
 
-          //skip element if is from a linked file and setting is off
-          if (!receiveLinkedModels && @base["isRevitLinkedModel"] != null && bool.Parse(@base["isRevitLinkedModel"].ToString()))
-            continue;
+          if (shouldConvertAsDisplayable)
+            throw new Exception("Converting as Direct Shape");
```
Comment (teocomi) [inline ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs:271]: Is there a better way to handle this than throwing?

File: ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs [hunk 2]
```diff
@@ -277,9 +313,145 @@ private void DeleteObjects(IReceivedObjectIdMap<Base, Element> previousObjects,
...
         }
+
+        return obj;
+      }
+
+      // Retries a conversion using the direct mesh setting.
+      // Used in the case of failed conversions
+      ApplicationObject ConvertAsDisplayable(Base @base, Dictionary<string, string> displayableSettings)
+      {
+        converter.SetConverterSettings(displayableSettings);
```
Comment (teocomi) [inline ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs:347]: Not sure it's a good UX to change this on their behalf, I'd rather just pass that in the report instead.

File: ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs [hunk 3]
```diff
@@ -202,68 +241,65 @@ private void DeleteObjects(IReceivedObjectIdMap<Base, Element> previousObjects,
...
-      conversionProgressDict["Conversion"] = 1;
-
-      // Get setting to skip linked model elements if necessary
-      var receiveLinkedModelsSetting = CurrentSettings.FirstOrDefault(x => x.Slug == "linkedmodels-receive") as CheckBoxSetting;
-      var receiveLinkedModels = receiveLinkedModelsSetting != null ? receiveLinkedModelsSetting.IsChecked : false;
-
-      var index = -1;
-      while (++index < Preview.Count)
+      // Converts an object based on direct mesh settings and conversion result
+      ApplicationObject ConvertObject(ApplicationObject obj, Base @base, Dictionary<string, string> displayableSettings)
```
Comment (teocomi) [inline ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs:251]: Let's move local functions out

File: ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs (hunk 2 above)
Comment (teocomi) [reply ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit.Receive.cs:347]: EDIT: let's add a new binding

**Review by AlanRynne — CHANGES_REQUESTED** [review]
Switching this to "requested changes" not because there's any changes pending on our side, but because I don't want to merge this without @connorivy express consent, as we've had to fix some rather complex conflicts on our side.

**Review by connorivy — COMMENTED** [review]
Made a small tweak and now I'm happy with the merge resolution

**Review by connorivy — APPROVED** [review]

**Review by AlanRynne — APPROVED** [review]
Amazing! Just browsed through your commit and looks great @connorivy.

Approving too in order to merge in 🥳

File: ConnectorRhino/ConnectorRhino/ConnectorRhinoShared/UI/ConnectorBindingsRhino.Receive.cs [hunk 4]
```diff
@@ -346,7 +346,7 @@ ApplicationObject NewAppObj()
       }
 
       //Handle objects convertable using displayValues
-      var fallbackMember = current["displayValue"] ?? current["@displayValue"];
+      var fallbackMember = DefaultTraversal.displayValuePropAliases.Where(o => current[o] != null)?.First();
```
[the remark this answers is not in the document]
Comment (AlanRynne) [reply ConnectorRhino/ConnectorRhino/ConnectorRhinoShared/UI/ConnectorBindingsRhino.Receive.cs]: Done!

**Pull request merged**